### PR TITLE
 Simplify broadcastable a bit further and add docs

### PR DESCRIPTION
**Pull request by mbauman**
Amazing what a great exercise writing documentation is. This is a followup to #26435, and more clearly lays out the requirements for defining a custom (non-AbstractArray) type that can participate in broadcasting.  The biggest functional change here is that types that define their own BroadcastStyle must now also implement `broadcastable` to be a no-op.

### CODE REVIEW

File: doc/src/manual/interfaces.md [hunk 1]
```diff
@@ -452,21 +453,43 @@ V = view(A, [1,2,4], :)   # is not strided, as the spacing between rows is not f
...
+### Broadcast Styles
+
+`Base.BroadcastStyle` is the abstract type from which all styles are
 derived. When used as a function it has two possible forms,
 unary (single-argument) and binary.
 The unary variant states that you intend to
 implement specific broadcasting behavior and/or output type,
-and do not wish to rely on the default fallback ([`Broadcast.Scalar`](@ref) or [`Broadcast.DefaultArrayStyle`](@ref)).
-To achieve this, you can define a custom `BroadcastStyle` for your object:
+and do not wish to rely on the default fallback ([`Broadcast.DefaultArrayStyle`](@ref))).
```
Comment (fredrikekre) [inline doc/src/manual/interfaces.md]: One too many `)`s

File: doc/src/manual/interfaces.md [hunk 2]
```diff
@@ -444,6 +444,7 @@ V = view(A, [1,2,4], :)   # is not strided, as the spacing between rows is not f
 | **Optional methods** | | |
 | `Base.BroadcastStyle(::Style1, ::Style2) = Style12()` | Precedence rules for mixing styles |
 | `Base.broadcast_indices(::StyleA, A)` | Declaration of the indices of `A` for broadcasting purposes (for AbstractArrays, defaults to `axes(A)`) |
+| `Base.broadcastable(x)` | Convert `x` to an object that has [`axes`] and supports indexing |
```
Comment (fredrikekre) [inline doc/src/manual/interfaces.md]: ``[`axes`](@ref) `` or just `` `axes` ``

File: doc/src/manual/interfaces.md [hunk 3]
```diff
@@ -452,21 +453,43 @@ V = view(A, [1,2,4], :)   # is not strided, as the spacing between rows is not f
...
+`collect(Base.broadcastable(x)) == collect(x)`. A notable exception are `AbstractString`s;
+they are special-cased to behave as scalars for the purposes of broadcast even though they
+are iterable collections of their characters.
+
+The next two steps (selecting the output array and implementation) are dependent upon
+determining a single answer for a given set of arguments. Broadcast must take all the varied
+types of its arguments and collapse them down to just one output array and one
+implementation. Broadcast calls this single answer a "style." Every broadcastable object
+each has its own preferred style, and a promotion-like system is used to combine these
+styles into a single answer — the "desination style".
```
Comment (fredrikekre) [inline doc/src/manual/interfaces.md]: `desination  ` -> `destination`

**Review by fredrikekre — COMMENTED** [review]
CI failure is
```
LoadError("sysimg.jl", 273, LoadError("broadcast.jl", 11, LoadError("broadcast.jl", 11, UndefVarError(Symbol("@__dot__")))))
```

**Comment by stevengj** [comment]
It seems like the `BroadcastStyle(x)` should be based on the value `x`, not on a type `BroadcastStyle(T)`.  (Imagine something like `PyObject` where the behavior depends on the runtime value.)   Is there a technical reason why it must take a type argument?

>The biggest functional change here is that types that define their own BroadcastStyle must now also implement `broadcastable` to be a no-op.

What is wrong with something like
```jl
broadcastable(x) = _broadcastable(x, BroadcastStyle(typeof(x))
_broadcastable(x, ::Unknown) = ...deprecation...
_broadcastable(x, s) = x
```
so that no-op is the default for types that define their own `BroadcastStyle`?

**Comment by mbauman** [comment]
This started as a documentation issue and morphed into the code changes that were required to make the documentation simple and understandable.

Yes, I agree the need for the type to support `ndims` is rather limiting.  I'd like to address that in the future — it'd be nice to just use `length(axes(arg))`.  Note, though, that it is required to be type-stable for a type-stable broadcast.

As far as your other suggestion, the issue with doing it that way is that then every type that wants to participate in broadcasting must _both_ define `broadcastable` (to avoid the collect fallback) and define their preferred `BroadcastStyle` (to avoid being classified as `Unknown`).  How I was using `Unknown` before was kind-of an abuse — there isn't a base implementation for broadcasting `Unknown` styles.  Indeed, the barebones `Foo26601` would fail in surprising ways with that implementation.  It's far more truthful to state that all types that quack like an array are `DefaultArrayStyle`.  We just don't know if they'll quack like an array until after they become `broadcastable`.

IMO: `broadcastable` is the "simpler" API.  It's the gateway into overriding behaviors.  `BroadcastStyle` is _far_ more complicated.  It makes more sense to make the more complicated API require coordination with the simpler one than the other way around.

File: doc/src/manual/interfaces.md (hunk 3 above)
Comment (mbauman) [reply doc/src/manual/interfaces.md]: Thanks!  (On all counts)

**Comment by stevengj** [comment]
> Note, though, that it is required to be type-stable for a type-stable broadcast.

Broadcasting a `PyObject` will basically never be type-stable, and that's okay.

In the type-stable case, you don't lose anything by making `BroadcastStyle(x)` take the object `x` rather than the type `T`, so you only gain flexibility this way.

**Comment by stevengj** [comment]
It’s disruptive for people using an unreleased 0.7 feature... isn’t this precisely the time to fix the API?

**Comment by stevengj** [comment]
Besides, it could be backwards compatible if the instance-based constructor were defined as lowercase broadcaststyle(x). This would be more consistent with most of our traits functions anyway.

**Comment by mbauman** [comment]
Yeah, we're in agreement here.  Are you saying you want me to do that in this PR?  I'd still prefer to do it separately.

**Comment by stevengj** [comment]
Separately is fine with me.

**Pull request merged**